ambry-api/src/main/java/com.github.ambry/router/ByteRange.java

```diff
@@ -0,0 +1,239 @@
+    }
+  }
+
+  /**
+   * Get the end offset for this range.
+   * @return The inclusive end offset for this range.
+   * @throws UnsupportedOperationException if the range does not have a defined start offset
+   *                                       (i.e. not of the type {@link ByteRangeType#OFFSET_RANGE})
+   */
+  public long getEndOffset() {
+    switch (getType()) {
+      case OFFSET_RANGE:
+        return endOffset;
+      default:
+        throw new UnsupportedOperationException("Cannot get end offset for range type: " + type);
+    }
+  }
+
+  /**
+   * Get the number of bytes to read from the end of an object.
+   * @return The number of bytes to read from the end of the object.
+   * @throws UnsupportedOperationException if the range is not of the type {@link ByteRangeType#LAST_N_BYTES})
+   */
+  public long getLastNBytes() {
+    switch (getType()) {
+      case LAST_N_BYTES:
+        return startOffset;
+      default:
+        throw new UnsupportedOperationException("Cannot get last N bytes for range type: " + type);
+    }
+  }
+
+  /**
+   * Get the {@link ByteRangeType} for the range.
+   * @return the {@link ByteRangeType} for the range.
+   */
+  public ByteRangeType getType() {
+    return type;
+  }
+
+  /**
+   * @return the size of the range, in bytes.
+   * @throws UnsupportedOperationException for {@link ByteRangeType#FROM_START_OFFSET} type ranges.
+   */
+  public long getRangeSize() {
+    switch (getType()) {
+      case OFFSET_RANGE:
+        return getEndOffset() - getStartOffset() + 1;
+      case LAST_N_BYTES:
+        return getLastNBytes();
+      default:
+        throw new UnsupportedOperationException("Cannot determine range size for range type: " + type);
+    }
+  }
+
+  /**
+   * Given the total size of a blob, generate a new {@link ByteRange} of type {@link ByteRangeType#OFFSET_RANGE} with
+   * defined start and end offsets that are verified to be within the supplied total blob size.
+   * @param totalSize the total size of the blob that this range corresponds to.
+   * @return the {@link ByteRange} with start and end offsets, or {@code null} if {@code this} is not within
+   *         the provided total blob size.
+   */
+  public ByteRange toResolvedByteRange(long totalSize) {
+    switch (getType()) {
+      case LAST_N_BYTES:
+        if (getLastNBytes() <= totalSize) {
+          return new ByteRange(totalSize - getLastNBytes(), totalSize - 1, ByteRangeType.OFFSET_RANGE);
+        }
+        break;
+      case FROM_START_OFFSET:
+        if (getStartOffset() < totalSize) {
+          return new ByteRange(getStartOffset(), totalSize - 1, ByteRangeType.OFFSET_RANGE);
+        }
+        break;
+      case OFFSET_RANGE:
+        if (getEndOffset() < totalSize) {
+          return new ByteRange(getStartOffset(), getEndOffset(), ByteRangeType.OFFSET_RANGE);
+        }
+        break;
+    }
+    return null;
```

Comment:
I can make this throw an exception.  This will keep it consistent with the rest of `ByteRange`, which typically doesn't return null.